art_internal_impl.hpp

```diff
@@ -94,78 +77,203 @@ basic_art_key<std::uint64_t>::make_external(std::uint64_t k) noexcept {
+  // to a method comparing a caller's key suffix against the key
+  // suffix stored in the leaf.
+  [[nodiscard, gnu::pure]] constexpr auto cmp(art_key_type k) const noexcept {
+    return k.cmp(get_key_view());
+  }
+
+  /// Return a view onto the value stored in the leaf.
   [[nodiscard, gnu::pure]] constexpr auto get_value_view() const noexcept {
-    return value_view{&value_start[0], value_size};
+    return value_view{data + key_size, value_size};
   }
 
 #ifdef UNODB_DETAIL_WITH_STATS
 
+  /// Return the byte size of the leaf data structure.
   [[nodiscard, gnu::pure]] constexpr auto get_size() const noexcept {
-    return compute_size(value_size);
+    return compute_size(key_size, value_size);
   }
 
 #endif  // UNODB_DETAIL_WITH_STATS
 
   [[gnu::cold]] UNODB_DETAIL_NOINLINE void dump(std::ostream &os,
                                                 bool /*recursive*/) const {
-    os << ", " << get_key() << ", value size: " << value_size << '\n';
+    os << ", ";
+    dump_key(os, get_key_view());
+    os << ", ";
+    dump_val(os, get_value_view());
+    os << '\n';
   }
 
+  /// Compute the required byte size of the leaf to hold the header,
+  /// key, and value.
+  ///
+  /// @param key_size The size in bytes of the portion of the key
+  /// stored in the leaf.
+  ///
+  /// @param val_size The size in bytes of the the value stored in the
+  /// leaf.
   [[nodiscard, gnu::const]] static constexpr auto compute_size(
-      value_size_type val_size) noexcept {
-    return sizeof(basic_leaf<Header>) + val_size - 1;
+      key_size_type key_size, value_size_type val_size) noexcept {
+    return sizeof(basic_leaf<Key, Header>) + key_size + val_size -
+           1  // because of the [1] byte on the end of the struct.
+        ;
   }
 
-  static constexpr std::size_t max_value_size =
-      std::numeric_limits<value_size_type>::max();
-
  private:
-  const art_key key;
+  /// The byte length of the key.
+  const key_size_type key_size;
+  /// The byte length of the value.
   const value_size_type value_size;
+  /// The leaf's key and value data starts at data[0].  The key comes
+  /// first followed by the data.
+  //
   // NOLINTNEXTLINE(modernize-avoid-c-arrays)
-  std::byte value_start[1];
+  std::byte data[1];
 };  // class basic_leaf
 
-template <class Header, class Db>
-[[nodiscard]] auto make_db_leaf_ptr(art_key k, value_view v,
-                                    Db &db UNODB_DETAIL_LIFETIMEBOUND) {
-  using leaf_type = basic_leaf<Header>;
+/// Return a unique pointer for a new leaf initialized with the
+/// caller's key and value.
+template <class Key, class Header, template <class> class Db>
+[[nodiscard]] auto make_db_leaf_ptr(basic_art_key<Key> k, value_view v,
+                                    Db<Key> &db UNODB_DETAIL_LIFETIMEBOUND) {
+  using leaf_type = basic_leaf<Key, Header>;
+
+  // TODO(thompsonbry) We should have a discussion about limits.  To
```

Comment:
@laurynas-biveinis Just FYI.  I kept things as they are, but let's discuss at some point.  It could also make sense to have an explicit type for the key size and the value size to we can freely set the designed boundary conditions.

Reply:
Sure